--- portmap_5beta/from_local.c~	2006-01-30 15:35:51.000000000 +0100
+++ portmap_5beta/from_local.c	2006-01-30 16:43:50.000000000 +0100
@@ -68,6 +68,8 @@
 static int num_local;
 static int num_addrs;
 static struct in_addr *addrs;
+extern struct in_addr bindaddr;
+extern int have_bindaddr;
 
 /* grow_addrs - extend list of local interface addresses */
 
@@ -114,6 +116,16 @@
      * presence of non-IP interfaces.
      */
 
+    /* if we have bindaddr it must be local, add it unconditionally
+     * in case of ioctl problems
+     * */
+    if (have_bindaddr && addrs == 0) {
+	if (num_local >= num_addrs)
+	    grow_addrs();
+	if (num_local < num_addrs)
+	    addrs[num_local++] = &bindaddr;
+    }
+
     if ((sock = socket(PF_INET, SOCK_DGRAM, 0)) < 0) {
 	perror("socket");
 	return (0);
