--- portmap_5beta/portmap.c.wiget	Fri Mar  7 19:45:46 2003
+++ portmap_5beta/portmap.c	Fri Mar  7 19:55:30 2003
@@ -95,8 +95,8 @@
 #ifdef SYSV40
 #include <netinet/in.h>
 #endif
-
-extern char *strerror();
+#include <errno.h>
+#include <string.h>
 #include <stdlib.h>
 
 #ifndef LOG_PERROR
@@ -124,7 +130,6 @@
 static void callit();
 struct pmaplist *pmaplist;
 int debugging = 0;
-extern int errno;
 
 #include "pmap_check.h"
 
